src/twisted/conch/interfaces.py

```diff
@@ -82,6 +82,26 @@ def getPty(term, windowSize, modes):
         fails, raise an exception.
         """
 
+    def setEnv(name, value):
```

Comment:
Unfortunately this isn't backwards compatible. [Interfaces are forever.](https://orbifold.xyz/interfaces-are-forever.html) You will have to create a new interface to host this method, e.g. `ISetEnv`, and check if the session object implements it in `request_env`.

Reply:
So, I understand that argument and I'm willing to do this, but I thought I'd first explain why I did it this way and see whether it changes your mind.

Generally speaking I try to fit into the style of existing code before almost any other consideration.  In this case, all of the `request_*` methods in `SSHSession` wrap their calls to the session object in very general exception handlers.  (I couldn't quite bring myself to imitate the bare `except`s that they use, but I imitated the rest of it.)  As such, it is effectively already true that `request_env` checks whether the session object implements the necessary method: if it doesn't, it results in an error being logged, but that's as far as it goes.  Of the current `ISession` implementers in Twisted itself, `docs/conch/examples/sshsimpleserver.py:ExampleSession`, `twisted.conch.manhole_ssh.TerminalSession`, and `twisted.conch.test.test_conch.RekeyAvatar` already fail to implement all of the methods in `ISession`, so for better or worse the current code can't be being strict about it.

Now, I can see that there might be an argument that `ISession` is a public interface and so external code is entitled to depend on it in other ways; but the above is why I took this approach in the first pass.

Reply:
> Now, I can see that there might be an argument that `ISession` is a public interface and so external code is entitled to depend on it in other ways; but the above is why I took this approach in the first pass.

Yeah, that's pretty much it. I don't view it so much as an argument as Twisted policy. But maybe this should be an exception if none of the real-world implementations actually _implement_ the interface. This seems like a question for the mailing list &mdash; again, I'm not too familiar with Conch.

Reply:
It's OK, I just wanted to get an idea of how firm the requirement was.  I've moved this to a new `ISessionSetEnv` interface (which necessitated some test rearrangements to be able to effectively test the case where the avatar doesn't implement that interface).